datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -121,58 +291,220 @@ where
+            )
+        }
+        _ => {
+            internal_err!("Unsupported 0 argument count for function to_timestamp_millis")
+        }
+    }
 }
 
 /// to_timestamp_micros SQL function
 pub fn to_timestamp_micros(args: &[ColumnarValue]) -> Result<ColumnarValue> {
-    handle::<TimestampMicrosecondType, _, TimestampMicrosecondType>(
-        args,
-        |s| string_to_timestamp_nanos_shim(s).map(|n| n / 1_000),
-        "to_timestamp_micros",
-    )
+    match args.len() {
+        1 => handle::<TimestampMicrosecondType, _, TimestampMicrosecondType>(
+            args,
+            |s| string_to_timestamp_nanos_shim(s).map(|n| n / 1_000),
+            "to_timestamp_micros",
+        ),
+        n if n >= 2 => {
+            handle_multiple::<TimestampMicrosecondType, _, TimestampMicrosecondType, _>(
+                args,
+                string_to_timestamp_nanos_with_format_shim,
+                |n| n / 1_000,
+                "to_timestamp_micros",
+            )
+        }
+        _ => {
+            internal_err!("Unsupported 0 argument count for function to_timestamp_micros")
+        }
+    }
 }
 
 /// to_timestamp_nanos SQL function
 pub fn to_timestamp_nanos(args: &[ColumnarValue]) -> Result<ColumnarValue> {
-    handle::<TimestampNanosecondType, _, TimestampNanosecondType>(
-        args,
-        string_to_timestamp_nanos_shim,
-        "to_timestamp_nanos",
-    )
+    match args.len() {
+        1 => handle::<TimestampNanosecondType, _, TimestampNanosecondType>(
+            args,
+            string_to_timestamp_nanos_shim,
+            "to_timestamp_nanos",
+        ),
+        n if n >= 2 => {
+            handle_multiple::<TimestampNanosecondType, _, TimestampNanosecondType, _>(
+                args,
+                string_to_timestamp_nanos_with_format_shim,
+                |n| n,
+                "to_timestamp_nanos",
+            )
+        }
+        _ => {
+            internal_err!("Unsupported 0 argument count for function to_timestamp_nanos")
+        }
+    }
 }
 
 /// to_timestamp_seconds SQL function
 pub fn to_timestamp_seconds(args: &[ColumnarValue]) -> Result<ColumnarValue> {
-    handle::<TimestampSecondType, _, TimestampSecondType>(
-        args,
-        |s| string_to_timestamp_nanos_shim(s).map(|n| n / 1_000_000_000),
-        "to_timestamp_seconds",
-    )
+    match args.len() {
+        1 => handle::<TimestampSecondType, _, TimestampSecondType>(
+            args,
+            |s| string_to_timestamp_nanos_shim(s).map(|n| n / 1_000_000_000),
+            "to_timestamp_seconds",
+        ),
+        n if n >= 2 => handle_multiple::<TimestampSecondType, _, TimestampSecondType, _>(
+            args,
+            string_to_timestamp_nanos_with_format_shim,
+            |n| n / 1_000_000_000,
+            "to_timestamp_seconds",
+        ),
+        _ => internal_err!(
+            "Unsupported 0 argument count for function to_timestamp_seconds"
+        ),
+    }
```

Comment:
These are also repeated, and failry generic-friendly it seems. Should be possible to extract a common 
```rust
fn to_timestamp_impl<T: ArrowTimestampType>(
    args: &[ColumnarValue],
    name: &str,
) -> Result<ColumnarValue> {
    let factor = match T::UNIT {
        TimeUnit::Second => 1_000_000_000,
        TimeUnit::Millisecond => 1_000_000,
        TimeUnit::Microsecond => 1_000,
        TimeUnit::Nanosecond => 1,
    };

    match args.len() {
        1 => handle::<T, _, T>(
            args,
            |s| string_to_timestamp_nanos_shim(s).map(|n| n / factor),
            name,
        ),
        n if n >= 2 => handle_multiple::<T, _, T, _>(
            args,
            string_to_timestamp_nanos_with_format_shim,
            |n| n / factor,
            name,
        ),
        _ => internal_err!("Unsupported 0 argument count for function {name}"),
    }
}
```

And then have the individual implementations use that, e.g. `to_timestamp_impl<TimestampSecondType>(args, "to_timestamp_seconds")` in the above case.

Reply:
That almost works - I needed to add + ScalarType<i64> to the signature to make it compile. I'll update the pull request with this improvement.